File: 02_paradigm_exploration/data_structure_manipulation/stack.scala
object StackFunctional {
  type Stack[A] = List[A]

  def push[A](stack: Stack[A], item: A): Stack[A] = item :: stack

  def pop[A](stack: Stack[A]): (A, Stack[A]) = stack match {
    case Nil       => throw new NoSuchElementException("Stack is empty")
    case x :: rest => (x, rest)
  }

  def peek[A](stack: Stack[A]): A = stack match {
    case Nil      => throw new NoSuchElementException("Stack is empty")
    case x :: _   => x
  }

  // Example usage
  def main(args: Array[String]): Unit = {
    val emptyStack: Stack[Int] = Nil
    val s1 = push(emptyStack, 42)
    val s2 = push(s1, 99)
    println("Peek: " + peek(s2))      // 99
    val (top, s3) = pop(s2)
    println("Popped: " + top)         // 99
    println("New Peek: " + peek(s3))  // 42
  }
}

File: 02_paradigm_exploration/data_structure_manipulation/stack.rb
class Stack
    def initialize
      @elements = []
    end
  
    def push(item)
      @elements.push(item)
    end
  
    def pop
      raise "Stack is empty" if empty?
      @elements.pop
    end
  
    def peek
      raise "Stack is empty" if empty?
      @elements.last
    end
  
    def empty?
      @elements.empty?
    end
  end
  
  # Example Usage
  stack = Stack.new
  stack.push(10)
  stack.push(20)
  puts stack.peek  # => 20
  puts stack.pop   # => 20
  puts stack.pop   # => 10
  